python-package/lightgbm/basic.py

```diff
@@ -2386,20 +2386,22 @@ def num_feature(self):
         else:
             raise LightGBMError("Cannot get num_feature before construct dataset")
 
-    def feature_num_bin(self, feature: int) -> int:
+    def feature_num_bin(self, feature: Union[int, str]) -> int:
         """Get the number of bins for a feature.
 
         Parameters
         ----------
-        feature : int
-            Index of the feature.
+        feature : int or str
+            Index or name of the feature.
 
         Returns
         -------
         number_of_bins : int
             The number of constructed bins for the feature in the Dataset.
         """
         if self.handle is not None:
+            if isinstance(feature, str):
+                feature = self.feature_name.index(feature)
```

Comment:
What if I want to use default feature names assigned on cpp side (`['Column_0', 'Column_1', ...]`)?

Also, feature names might be overwritten at cpp side:
https://github.com/microsoft/LightGBM/blob/248fbfa6e8fbf3c284a713681eed4d988e16131d/src/io/dataset.cpp#L1465-L1479
```python
import numpy as np
import lightgbm as lgb

X = np.random.random((50, 5))
y = np.random.random((50))

feature_names=[f'x{i}' for i in range(X.shape[1])]
lgb_data1 = lgb.Dataset(X, y, feature_name=feature_names).construct()
lgb_data2 = lgb.Dataset(X, y, feature_name=feature_names).construct()
lgb_data1.add_features_from(lgb_data2)
lgb_data1.add_features_from(lgb_data2)

bst = lgb.train({}, lgb_data1, num_boost_round=10)

print(bst.dump_model()['feature_names'])
#> ['x0', 'x1', 'x2', 'x3', 'x4', 'D2_x0', 'D2_x1', 'D2_x2', 'D2_x3', 'D2_x4', 'D3_x0', 'D3_x1', 'D3_x2', 'D3_x3', 'D3_x4']
```

I believe we should use the same code as here to handle the above mentioned cases
https://github.com/microsoft/LightGBM/blob/5631366175f68a28bd174982cca16975f7fb48af/python-package/lightgbm/basic.py#L3784-L3785

Reply:
The `add_features_from` method updates the feature names https://github.com/microsoft/LightGBM/blob/248fbfa6e8fbf3c284a713681eed4d988e16131d/python-package/lightgbm/basic.py#L2523 so in the example you posted `lgb_data1.feature_num_bin('D3_x4')` works.

When not using feature names in the constructor I don't think people can expect that getting the number of bins by name works, or even know what column names are assigned. I could replace `self.feature_name` with `self.get_feature_name()` when looking for the feature names but I'm not sure if that use case is worth it.

Reply:
I think it's not a big deal to replace `self.feature_name` with `self.get_feature_name()` for the sake of increasing the coverage of function use cases. Are you confused by potential performance degradation or what?

Reply:
> Do you think maybe we can try to keep the feature names in sync with the cpp side?

I think it's great idea!
Seems it will be enough to just set `self.feature_name = self.get_feature_name()` after Dataset construction.

Reply:
added in 87b57ca6558bd42b1aea8b437fae18dad15f1411